src/events/google_cloud.js

```diff
@@ -79,6 +87,132 @@ function bigQueryWrapper(wrappedFunction) {
+                            }
+                        } else if (config.method === 'createTopic') {
+                            callbackResponse.topic = arg2;
+                        }
+                    }
+                    finalizeEvent(event, startTime, err, callbackResponse);
+                    resolve();
+                    if (callback) {
+                        if (arg3) {
+                            callback(err, arg2, ...arg3);
+                        } else {
+                            callback(err, arg2);
+                        }
+                    }
+                };
+            });
+            tracer.addEvent(event, responsePromise);
+        } catch (err) {
+            tracer.addException(err);
+        }
+        return original.apply(this, [config, patchedCallback]);
+    };
+}
+
+/**
+ * Wrap pubsub request function.
+ * @param {Function} original pubsub request function.
+ * @returns {Function} The wrapped function
+ */
+function wrapPubSubPullFunction(original) {
+    return function internalPubSubPullFunction(request, options, callback) {
+        try {
+            let pubsubProjectId;
+            if (request && request.subscription) {
+                const subscriptionSplited = request.subscription.split('/');
+                if (subscriptionSplited.length > 1) {
+                    [, pubsubProjectId] = subscriptionSplited;
+                }
+            }
+            const { event, startTime } = initializeEvent(
+                GOOGLE_CLOUD_TYPES.pubsub.name, pubsubProjectId, 'Pull'
+            );
+            const patchedCallback = (err, res) => {
+                const callbackResponse = {};
+                if (res && res.receivedMessages) {
+                    const messageIds = res.receivedMessages.reduce((acc, current) => {
+                        acc.push(current.message.messageId);
+                        return acc;
+                    }, []);
+                    callbackResponse.messageIds = messageIds;
+                }
+                finalizeEvent(event, startTime, err, callbackResponse);
+            };
+            let responsePromise;
+            // in case callback has given from the client.
+            if (callback) {
+                responsePromise = original.apply(this, [request, options, patchedCallback]);
+                tracer.addEvent(event, responsePromise);
+            }
+            responsePromise = original.apply(this, [request, options, callback]);
+            tracer.addEvent(event, responsePromise);
+            return responsePromise.then((res) => {
+                const [response] = res;
+                patchedCallback(null, response);
+                return res;
+            }, (err) => {
+                patchedCallback(err, null);
+                throw err;
+            });
+        } catch (err) {
+            tracer.addException(err);
+        }
+        return original.apply(this, [request, options, callback]);
```

Comment:
when are you reaching this line? I feel like you might be calling the original function at least twice

Reply:
you right, fixed.